sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -540,22 +546,39 @@ public virtual ServiceBusReceivedMessage AmqpMessageToSBMessage(AmqpMessage amqp
                 }
             }
 
+            return annotatedMessage;
+        }
+
+        public virtual ServiceBusReceivedMessage AmqpMessageToSBReceivedMessage(AmqpMessage amqpMessage, bool isPeeked = false)
+        {
+            AmqpAnnotatedMessage annotatedMessage = AmqpMessageToAnnotatedMessage(amqpMessage, isPeeked);
+
+            ServiceBusReceivedMessage sbMessage = new ServiceBusReceivedMessage(annotatedMessage);
+
             // lock token
 
-            if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
+            sbMessage.LockTokenGuid = ParseGuidBytes(amqpMessage.DeliveryTag);
+
+            amqpMessage.Dispose();
+
+            return sbMessage;
+        }
+
+        public virtual Guid ParseGuidBytes(ReadOnlyMemory<byte> bytes)
+        {
+            if (bytes.Length == GuidSizeInBytes)
             {
-                Span<byte> guidBytes = stackalloc byte[GuidSizeInBytes];
-                amqpMessage.DeliveryTag.AsSpan().CopyTo(guidBytes);
-                if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
+                // Use TryRead to avoid allocating an array if we are on a little endian machine.
+                if (!BitConverter.IsLittleEndian || !MemoryMarshal.TryRead<Guid>(bytes.Span, out var lockTokenGuid))
```

Comment:
But the first bool expression will almost always be true (as most machines are little endian), and so it will execute the "slow path". If we want to make the mainline scenario fast (not allocating), you just need to copy the code from here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Guid.cs#L63

Reply:
I don't follow - if most machines are little endian than the first expression would be false.

Reply:
I guess I missed the '!'. It was late in the day :-)

Reply:
All good 😄